tools/codegen/api/dispatcher.py

```diff
@@ -74,18 +74,23 @@ def arguments(func: FunctionSchema) -> Sequence[DispatcherArgument]:
 
 # Given a set of CppArguments in scope, return a sequence of dispatcher
 # expressions that translate the cpp API into dispatcher API
-def cppargument_exprs(a: CppArgument, *, tensor_options: Optional[CppArgument]) -> Sequence[DispatcherExpr]:
+def cppargument_exprs(a: CppArgument, *, tensor_options: Optional[CppArgument], process_tensoroptions: Optional[str]) -> Sequence[DispatcherExpr]:
     if isinstance(a.argument, TensorOptionsArguments):
-        if local.use_c10_dispatcher() is UseC10Dispatcher.full:
+        if process_tensoroptions == 'scatter':
             ta = a.argument
             return [
                 DispatcherExpr(type=argument_type(ta.dtype), expr=f'optTypeMetaToScalarType({a.name}.dtype_opt())'),
                 DispatcherExpr(type=argument_type(ta.layout), expr=f'{a.name}.layout_opt()'),
                 DispatcherExpr(type=argument_type(ta.device), expr=f'{a.name}.device_opt()'),
                 DispatcherExpr(type=argument_type(ta.pin_memory), expr=f'{a.name}.pinned_memory_opt()'),  # weird discrep
             ]
+        elif process_tensoroptions == 'gather':
+            return [DispatcherExpr(type='const TensorOptions &', expr="TensorOptions().dtype(dtype).layout(layout).device(device).pinned_memory(pin_memory)")]
         else:
+            assert process_tensoroptions is None
```

Comment:
It's inconsistent that we assert on this but not the same for `tensor_options`, I'd do both or neither. (Actually, best would be to bundle them into a little named tuple - correct by construction!)

Reply:
this assertion is just to make the switch/case comprehensive and make sure we didn't miss a case. The `tensor_options` argument is mostly independent from that.